pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/segment/SegmentFlushThresholdComputer.java

```diff
@@ -157,13 +182,25 @@ public int computeThreshold(StreamConfig streamConfig, CommittingSegmentDescript
       }
     }
     targetSegmentNumRows = capNumRowsIfOverflow(targetSegmentNumRows);
+    targetSegmentNumRows = applySegmentSizeVariation(segmentSizeVariation, targetSegmentNumRows);
     SegmentSizeBasedFlushThresholdUpdater.LOGGER.info(
         "Committing segment size {}, current ratio {}, setting threshold for {} as {}",
         committingSegmentSizeBytes, _latestSegmentRowsToSizeRatio, newSegmentName, targetSegmentNumRows);
 
     return (int) targetSegmentNumRows;
   }
 
+  private long applySegmentSizeVariation(double segmentSizeVariation, long targetSegmentNumRows) {
+    if (segmentSizeVariation > 0.0 && segmentSizeVariation < 1.0) {
+      double variation = (1 - segmentSizeVariation) + 2 * segmentSizeVariation * RANDOM.nextDouble();
+      // Unnecessary to check for negative variation as it qwill always be positive
+      if (variation > 0) {
+        return (long) (targetSegmentNumRows * variation);
+      }
+    }
+    return targetSegmentNumRows;
```

Comment:
We probably want a lower bound of this to avoid tiny segments.
There is side effect of this which can cause segment size tuning hard to converge. Should we limit the variation to some relatively small value? 100% variation can cause quite unstable behavior

Reply:
set to 0-0.5